### PR TITLE
Pytest fixes and reactivate precommit prospector

### LINKED ISSUE #46: Reactivate precommit prospector
Some errors come from the pytests because of apparently unused imports, due to the way pytest works.

I suggest to first fix the problematic tests in #45 and then work on this issue.

### PR DESCRIPTION
**Pull request by normarivano**
Some errors came from the pytests because of unused imports, unused arguments, too many variables and similar issues related to the way pytest works. We fix here the remaining errors and we reactivate the precommit prospector. 

Fixes #46

### CODE REVIEW

File: aiida_wannier90/calculations.py [hunk 1]
```diff
@@ -176,9 +172,9 @@ def prepare_for_submission(self, folder):
             param_dict.update({'postproc_setup': True})
 
         try:
-            local_input_folder = self.inputs.local_input_folder
+            local_input_folder = self.inputs.local_input_folder  #pylint: disable=unused-variable
```
Comment (greschd) [inline aiida_wannier90/calculations.py]: It seems to me this variable actually isn't used in the rest of the function.

File: aiida_wannier90/io/_group_list.py [hunk 2]
```diff
@@ -19,10 +19,9 @@ def group_list(values):
         else:
             groups.append(sorted(set([current_start, v1])))
             current_start = v2
-    # final group
-    else:
+        # final group
         groups.append(sorted(set([current_start, v2])))
```
Comment (greschd) [inline aiida_wannier90/io/_group_list.py]: Here the ``groups.append`` should happen only after the for-loop has finished -- that is, both ``groups.append`` and ``return groups`` should be indented one more level to the left. Sorry, my previous comment wasn't clear on this point.

File: tests/test_local_input.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
+# -*- coding: utf-8 -*-
```
Comment (greschd) [inline tests/test_local_input.py]: Isn't this the same file as ``tests/calculations/test_local_input``?

File: tests/test_caching.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
```
Comment (greschd) [inline tests/test_caching.py]: In #57, I intentionally deleted these tests - the new tests don't need to explicitly run Wannier90. Also, it's not really aiida-wannier90's job to test the AiiDA caching feature - I think I just put this here initially because I was working on both in parallel.

File: aiida_wannier90/__init__.py [hunk 5]
```diff
@@ -16,9 +20,9 @@
 ## If upgraded, remember to change it also in setup.json (for pip)
 __version__ = "2.0.0a1"
 
-from . import io
-from . import parsers
-from . import orbitals
+#from . import io
```
Comment (greschd) [inline aiida_wannier90/__init__.py]: What's the reason for not importing the submodules here? Too much startup time?

In any case: If we remove it, I think the code should be deleted, not just commented out.

File: aiida_wannier90/io/__init__.py [hunk 6]
```diff
@@ -7,4 +7,4 @@
 This submodule contains helper functions to create input files.
 """
 
-from ._write_win import write_win
+#from ._write_win import write_win
```
Comment (greschd) [inline aiida_wannier90/io/__init__.py]: What's the reason for this change? Note that this changes the public API: previously, the helper function was publicly accessible as ``aiida_wannier90.io.write_win``. Now it would be an internal - only function.

File: examples/workflows/run_w90.py [hunk 7]
```diff
@@ -3,14 +3,12 @@
...
-import argparse
-from aiida.common.exceptions import NotExistent
 from aiida.orm import Str
 from aiida.orm import Dict
-from aiida.orm.nodes.data.structure import StructureData
+from aiida.orm import load_node, Code
 from aiida.orm.nodes.data.array.kpoints import KpointsData
-from aiida.engine import run, submit
-from aiida_wannier90.workflows.W90 import SimpleWannier90WorkChain
+from aiida.engine import submit  # run unused
```
Comment (greschd) [inline examples/workflows/run_w90.py]: Here we can just remove the comment.

File: aiida_wannier90/calculations.py (hunk 1 above)
Comment (normarivano) [reply aiida_wannier90/calculations.py]: Concerning this point, we had a discussion with @giovannipizzi so I was not sure if we want to delete it or not.

File: aiida_wannier90/io/_group_list.py (hunk 2 above)
Comment (normarivano) [reply aiida_wannier90/io/_group_list.py]: Yes, sorry!

File: tests/test_caching.py (hunk 4 above)
Comment (normarivano) [reply tests/test_caching.py]: Same as above... my fault.

File: aiida_wannier90/__init__.py (hunk 5 above)
Comment (normarivano) [reply aiida_wannier90/__init__.py]: I commented them just to avoid complains about "unused imports" and I'm not sure if we need/want to keep it.

File: aiida_wannier90/calculations.py (hunk 1 above)
Comment (greschd) [reply aiida_wannier90/calculations.py]: It looks to me like the statement doesn't have any effect -- further down the ``local_input_folder`` is again obtained directly from ``self.inputs``.

File: aiida_wannier90/__init__.py (hunk 5 above)
Comment (greschd) [reply aiida_wannier90/__init__.py]: Ah, I see. The imports here are unused, but on purpose (generally, that's fine in ``__init__.py`` files). The difference this makes is that with the imports the following would work:

```python
import aiida_wannier90
aiida_wannier90.io.write_win(...)
```

whereas otherwise you would have to explicitly use
```python
import aiida_wannier90.io
```
or 
```python
from aiida_wannier90.io import write_win
```

Typically I tend to add these imports because it's a bit more convenient to use, but it has a cost in taking longer to initially load the module. That is important mostly if the package is imported somewhere in a command-line tool.

Here I'm fine either way, but it should be a conscious decision.

Tagging @giovannipizzi for comment.

Comment (normarivano) [reply aiida_wannier90/__init__.py]: I discussed with Giovanni and he agrees on keeping the imports and just disable the warnings. Otherwise he suggested to use   ```__all__``` but he was not sure about how it works. What do you think?

Comment (greschd) [reply aiida_wannier90/__init__.py]: Ok, yeah I agree there should be an ``__all__`` here. That might even solve the pylint warnings, but I'm not sure about that.

It should be
```python
__all__ = ("io", "parsers", "orbitals")
```
after the import statements. This determines what gets imported on ``from aiida_wannier90 import *``, and some tools (e.g. ipython) also take it into account for autocompletion.

Looking around the code, it seems we're missing ``__all__`` in a few places.. I'll make an issue for myself.

File: aiida_wannier90/__init__.py [hunk 8]
```diff
@@ -16,9 +20,9 @@
 ## If upgraded, remember to change it also in setup.json (for pip)
 __version__ = "2.0.0a1"
 
-from . import io
-from . import parsers
-from . import orbitals
+from . import io  # noqa: disable=F401
+from . import parsers  # noqa: disable=F401
+from . import orbitals  # noqa: disable=F401
 ## If this is imported then AiiDA ORM is loaded and this fails if the load_dbenv has not
```
Comment (greschd) [inline aiida_wannier90/__init__.py]: I think this comment is outdated -- previously `calculations` could not be imported without a loaded dbenv, but it appears this is no longer the case (there were some changes in `aiida-core` affecting this).

File: aiida_wannier90/parsers.py [hunk 9]
```diff
@@ -105,8 +105,10 @@ def parse(self, **kwargs):
         if exiting_in_stdout:
             return self.exit_codes.ERROR_EXITING_MESSAGE_IN_STDOUT
 
+        return self.exit_codes.ExitCode(0)
```
Comment (greschd) [inline aiida_wannier90/parsers.py]: I guess this is the change that causes the tests to fail.

Comment (normarivano) [reply aiida_wannier90/parsers.py]: I was getting the following error: 

_aiida_wannier90/parsers.py
  Line: 30
    pylint: inconsistent-return-statements / Either all return statements in a function should return an expression, or none of them should. (col 4)_

So, either I disable the pylint exception or I was looking a for trivial ExitCode . But maybe something change concerning this specific exit code...

File: aiida_wannier90/__init__.py (hunk 8 above)
Comment (normarivano) [reply aiida_wannier90/__init__.py]: Great!

File: aiida_wannier90/parsers.py (hunk 9 above)
Comment (greschd) [reply aiida_wannier90/parsers.py]: @sphuber can answer definitively, but I think in this case it's fine not to return anything.

To be honest I'm not too fond of this check, for all I care we could disable it globally as in https://github.com/aiidateam/aiida-quantumespresso/blob/develop/.pylintrc#L64

Comment (sphuber) [reply aiida_wannier90/parsers.py]: You can return an exit code with status 0 as that has the same effect as returning `None`, but it should be simply `return ExitCode()`. The default value is `0` but you can of course make it explicit `return ExitCode(0)`. The `self.exit_codes` is a property to a mapping of the defined exit codes on the calculation class. So you can only use the exit code labels used in their definition, e.g. if the `CalcJob` spec contains `spec.exit_code(418, 'ERROR_I_AM_A_TEAPOT') then you can get it with `self.exit_codes.ERROR_I_AM_A_TEAPOT`, which will simply return `ExitCode(418, 'ERROR_I_AM_A_TEAPOT')`.

Comment (sphuber) [reply aiida_wannier90/parsers.py]: But yes, I also typically just ignore this warning because I think in these cases I think it is fine to have either `None` or an `ExitCode` as return statement

Comment (normarivano) [reply aiida_wannier90/parsers.py]: Ok, thank´s both of you!
Maybe for the moment I can disable it locally and for the future we could do something similar to quantum espresso? @giovannipizzi  what do you think?

File: tests/conftest.py [hunk 10]
```diff
@@ -52,8 +51,10 @@ def fixture_folderdata():
 
     # TODO: wrap 'replacement_mapping in 'types.MappingProxyType' after Python2 support
     # is dropped, for immutability.
-    def _fixture_folderdata(dir_path, replacement_mapping={}):
+    def _fixture_folderdata(dir_path, replacement_mapping=None):
         # TODO: Remove cast to 'str' when Python2 support is dropped.
```
Comment (greschd) [inline tests/conftest.py]: This comment should be below, at the `dir_path` line.

File: aiida_wannier90/parsers.py (hunk 9 above)
Comment (greschd) [reply aiida_wannier90/parsers.py]: Yeah, let's just ignore it locally for now.

Any of the proposed solutions would be ok, really -- since this doesn't influence behavior or interface, it's not really important which one we choose.

Comment (giovannipizzi) [reply aiida_wannier90/parsers.py]: Ok to disable it, for sure for the file!

**Comment by greschd** [comment]
@giovannipizzi if you agree, as an admin you should have the power to merge.

**Comment by greschd** [comment]
The fix for the documentation problem is in #73, so we could just merge that first. There might be a small merge conflict in `doc/source/conf.py`, I will take care of that.
